x/oracle/keeper/icq.go

```diff
@@ -0,0 +1,45 @@
+package keeper
+
+import (
+	"time"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+
+	abcitypes "github.com/tendermint/tendermint/abci/types"
+
+	cerrs "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v6/modules/core/24-host"
+
+	"github.com/provenance-io/provenance/x/oracle/types"
+)
+
+// QueryOracle sends an ICQ to the other chain's module
+func (k Keeper) QueryOracle(ctx sdk.Context, query wasmtypes.RawContractMessage, channel string) (uint64, error) {
+	chanCap, found := k.scopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(k.GetPort(ctx), channel))
+	if !found {
+		return 0, cerrs.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
+	}
+
+	q := types.QueryOracleRequest{
+		Query: query,
+	}
+
+	reqs := []abcitypes.RequestQuery{
+		{
+			Path: "/provenance.oracle.v1.Query/Oracle",
+			Data: k.cdc.MustMarshal(&q),
+		},
+	}
+
+	timeoutTimestamp := ctx.BlockTime().Add(time.Minute).UnixNano()
```

Comment:
It surprises me that nano is used here. It also annoys me that they don't document anywhere that it should be nanoseconds. I had to dig down to where it's used; it's ultimately compared to `consensusState.GetTimestamp()` which is documented to be in nano.

I guess max uint64 as nanoseconds since epoch is still `2554-07-21` (a Sunday), so we've got more time than I expected.

(just musing, feel free to resolve this comment)